finality-aleph/src/data_io.rs

```diff
@@ -1,13 +1,189 @@
+        let mut timeout = Delay::new(Duration::from_millis(5000));
+        let mut import_stream = self.client.import_notification_stream();
+        loop {
+            tokio::select! {
+                Some(message) = &mut self.messages_rx.next() => {
+                    trace!(target: "afa", "Received message at Data Store {:?}", message);
+                    self.add_message(message);
+                }
+                Some(block) = &mut import_stream.next() => {
+                    trace!(target: "afa", "Block import notification at Data Store for block {:?}", block);
+                    self.add_block(block.hash);
+                }
+                _ = &mut timeout => {
+                    trace!(target: "afa", "Data Store timeout");
+                    let keys : Vec<_> = self.dependent_messages.keys().cloned().collect();
+                    for block_hash in keys {
+                        if self.client.header(BlockId::Hash(block_hash)).is_ok() {
+                            self.add_block(block_hash);
+                        }
+                    }
+                    timeout = Delay::new(Duration::from_millis(5000));
+                }
+                _ = &mut exit => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn forget_message(&mut self, message_id: MessageId) {
+        self.message_requirements.remove(&message_id);
+        if let Some(message) = self.pending_messages.remove(&message_id) {
+            for block_hash in message.included_data() {
+                if let Entry::Occupied(mut entry) = self.dependent_messages.entry(block_hash) {
+                    entry.get_mut().remove(&message_id);
+                    if entry.get().is_empty() {
+                        entry.remove_entry();
+                    }
+                }
+            }
+        }
+    }
+
+    fn add_pending_message(&mut self, message: AlephNetworkData<B>, requirements: Vec<B::Hash>) {
+        let message_id = self.next_message_id;
+        self.next_message_id += 1;
+        for block_hash in requirements.iter() {
+            self.dependent_messages
+                .entry(*block_hash)
+                .or_insert_with(HashSet::new)
+                .insert(message_id);
+        }
+        self.message_requirements
+            .insert(message_id, requirements.len());
+        self.pending_messages.insert(message_id, message);
+
+        if message_id >= PENDING_MESSAGE_LIFETIME {
+            self.forget_message(message_id - PENDING_MESSAGE_LIFETIME)
+        }
+    }
+
+    fn add_message(&mut self, message: AlephNetworkData<B>) {
+        let requirements: Vec<_> = message
+            .included_data()
+            .into_iter()
+            .filter(|b| {
+                if self.available_blocks.contains(b) {
+                    return false;
+                }
+                if self.client.header(BlockId::Hash(*b)).is_ok() {
+                    self.available_blocks.put(*b, ());
```

Comment:
it's not clear to me why to put directly instead of via `self.add_block`.

Reply:
There was none and I think `self.add_block` is even better here